### PR TITLE
feat(docs): add Website Testing decision record

### LINKED ISSUE #307: Decision: choose a website testing approach
# Context
As Emblem matures, we'll need to minimize the friction involved in contributing to Emblem.

A thorough _website ~backend~_ testing plan is valuable for several reasons:
1) We reduce (if not eliminate) `main`-breaking PRs
2) We speed up the development loop for new contributors (who can get feedback faster via automated tests, vs. human reviewers)
3) We can keep `main` in a known-good state, which allows contributors to focus on _their contributions_ instead of fixing regressions.

**April 2022 Note**
This issue incorporates @pattishin has done to reevaluate #47. Going forward, we'll be using a client-side framework called [Lit](https://lit.dev).

Accordingly, we **must** use a Browser or DOM-based testing framework, as opposed to pure HTTP ones (which only work well for server-side-rendered pages).

# Decisions

## Testing approach
At a high level, there are three types of website testing frameworks available:
- Generic web-driving frameworks (e.g. [Selenium](https://www.selenium.dev/) and [Puppeteer](https://github.com/puppeteer/puppeteer))
- Component-based testing frameworks (e.g. [Jest](https://jestjs.io/))
- End-to-end testing frameworks (e.g. [Cypress](https://www.cypress.io/) and [Playwright](https://playwright.dev))

The sections below compare the various framework types.

### Generic web-driving solutions

✅  Proven model understandable to most of our users ("bias towards boring")
✅  Idiomatic syntax (not much new our team needs to learn)
❌  Not built specifically for testing
❌  Sometimes flaky and unreliable (true for Selenium, per @engelke)

**Examples:** [Selenium](https://www.selenium.dev/) and [Puppeteer](https://github.com/puppeteer/puppeteer)

### End-to-end frameworks

✅  Built [specifically](https://www.cypress.io/how-it-works/) for end-to-end website testing
✅  Useful features (e.g. automatic screenshotting in Cypress) to make testing websites easier.
✅  Less flaky than _Generic Web-driving Solutions_.
❌  Library-specific syntax is fairly novel and would require a moderate (i.e. non-trivial) degree of learning

### Component-based frameworks

✅  Test specific components, so low likelihood of flakiness
❌  No standalone end-to-end testing support, judging by [this StackOverflow comment](https://stackoverflow.com/questions/66217682/should-i-use-both-cypress-and-jest-together#:~:text=They%20are%20orthogonal.,the%20choice%20is%20quite%20straightforward) and [these references to Puppeteer](https://jestjs.io/docs/puppeteer#use-jest-puppeteer-preset).

# Recommendations

## **Discard:** Generic web-driving solutions

These should arguably be considered a legacy approach to end-to-end tests, and were largely included (in this analysis) for completeness.

If we don't end up choosing them, we should clearly communicate _why_ we didn't to external developers. (**The answer:** we now have frameworks that are designed _specifically_ for E2E testing websites, and have more features targeted towards E2E testing.)

## **Adopt:** End-to-end testing framework

Per our (internal) testing strategy document, we want to focus on higher-level tests first. Since end-to-end _testing frameworks_ are purpose-built for end-to-end _tests_ and have a 'higher common denominator', we should stick to those frameworks.

### **Adopt:** Playwright

According to [this article](https://medium.com/sparebank1-digital/playwright-vs-cypress-1e127d9157bd), the two are relatively even for our purposes - with one major exception in favor of Playwright.

Playwright's "killer feature" is its ability to test cross-page (and cross-domain!) user flows. This is an important ingredient if we hope to test `auth`-related functionality (login/logout) in the future.

## **Pause:** component-based testing frameworks

We may want to adopt component-based testing frameworks (like [Jest](https://jestjs.io)) if and when we decide to start implementing per-component 'unit' tests for Emblem.

Right now (per the internal testing strategy doc), we are more focused on test _coverage_ (which lends itself to end-to-end tests) than test _granularity_ (which maps to unit tests).

### PR DESCRIPTION
**Pull request by ace-n**
Closes #307

### CODE REVIEW

File: docs/decisions/2022-04-website-testing.md [hunk 1]
```diff
@@ -0,0 +1,39 @@
+# Testing the Website
```
Comment (grayside) [inline docs/decisions/2022-04-website-testing.md:1]: Please use the [template](https://github.com/GoogleCloudPlatform/emblem/blob/main/docs/decisions/template.md)

File: docs/decisions/2022-04-website-testing.md [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+## Decisions
+
+### Testing Approach
+
+In light of the goals above, we decided as a team to focus on **End-to-End (E2E) tests** prior to Public Preview.
+
+### Tooling
+
+As a team, we evaluated [several different tools](https://modern-web.dev/docs/test-runner/browser-launchers/overview/). We [came to the conclusion](https://github.com/GoogleCloudPlatform/emblem/issues/307#issuecomment-1111508110) that [Playwright](https://playwright.dev) was the best fit for our needs.
```
Comment (grayside) [inline docs/decisions/2022-04-website-testing.md:33]: Please in-line notes about why Playwright, and any trade-offs or concerns to be on the lookout for. Are there any circumstances we anticipate as a higher risk for this dependency that could cause us to reconsider?

File: docs/decisions/2022-04-website-testing.md [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+
+## Revision Criteria
+
+We plan to revise this decision once Emblem enters Public Preview.
```
Comment (grayside) [inline docs/decisions/2022-04-website-testing.md:37]: Let's remove launch stages as a consideration here. The value in the decision record would be from the underlying requirements changes implied by these launch stages -- the readers won't gain insight from our launch labels.

If there aren't changes, then the criteria might be more a note on this being a trial period to see how this tool works for us.

**Review by grayside — CHANGES_REQUESTED** [review]

**Comment by ace-n** [comment]
@grayside thanks for the review - I think the right move here is to split **overall testing strategy** and **which framework to use** into separate PRs.

This one covers the former; I'll create a separate PR for the latter.

File: docs/decisions/2022-04-website-testing.md (hunk 1 above)
Comment (ace-n) [reply docs/decisions/2022-04-website-testing.md]: Done.

File: docs/decisions/2022-04-website-testing.md (hunk 2 above)
Comment (ace-n) [reply docs/decisions/2022-04-website-testing.md]: This was broken out into [its own PR] #366

File: docs/decisions/2022-04-website-testing-philosophy.md [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+
+_See the [Google SRE Guide](https://sre.google/sre-book/testing-reliability/#fig_testing_hierarchy) for an explanation of these test types._
+
+The Emblem Website can have as many of each test type as we'd like it to. Our only constraint is the time we'd need to take to implement each test.
+
+## Decisions
+
+For **Public Preview**, we should focus primarily on _a single test type_ to minimize time-to-value.
+
+To maximize our initial return on investment, our team decided to focus on **End-to-End (E2E) tests** prior to Public Preview.
```
Comment (grayside) [inline docs/decisions/2022-04-website-testing-philosophy.md:48]: I don't think "Public Preview" will be a useful milestone marker to most readers, as it lacks context.

Could we add some words in the direction of "In order to have test coverage before recommending use of Emblem (Public Preview)..."

File: docs/decisions/2022-04-website-testing-philosophy.md [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+
+Second, we don't expect our **initial** tests to simplify the contribution process.
+
+Third, we expect our initial tests to help us _detect_ - but **not debug** - problems.
+
+> The second and third issues can be solved by adding more **unit tests** and **integration tests** respectively. Those are things we may investigate later when we have more time.
+
+### Revisiting this Decision
+
+We plan to continue using Playwright so long as it continues meeting our project's needs.
```
Comment (grayside) [inline docs/decisions/2022-04-website-testing-philosophy.md:64]: Let's remove this section if we don't have something specific to say.

**Review by grayside — APPROVED** [review]

File: docs/decisions/2022-04-website-testing-philosophy.md (hunk 4 above)
Comment (ace-n) [reply docs/decisions/2022-04-website-testing-philosophy.md]: Dropped references to specific launch stages.

File: docs/decisions/2022-04-website-testing-philosophy.md (hunk 5 above)
Comment (ace-n) [reply docs/decisions/2022-04-website-testing-philosophy.md]: Done.

**Pull request merged**